### PR TITLE
Fix ZeRO parameter initialization for tensors with `requires_grad=True`

### PR DESCRIPTION
**Pull request by XuehaiPan**
This PR fixes a bug that occurs when the zero tensor ctor is called with `requires_grad=True`. `deepspeed` wraps the original tensor ctor (e.g., `torch.zeros()`) with a `tensor.to(dtype)` operation that makes the results become a non-leaf tensor with a `<ToCopyBackward>`.

In this PR, we manually call `.detach_()` to detach from the computation graph and then `.requires_grad_()` to re-enable the grad tracking mode. It is safe to detach the graph here because `zero_wrapper_for_fp_tensor_constructor` wraps only tensor constructor functions. The constructor functions always create leaf tensors.

```python
In [1]: import torch

In [2]: x = torch.zeros(4, requires_grad=True, device=torch.device('cuda'))

In [3]: x
Out[3]: tensor([0., 0., 0., 0.], device='cuda:0', requires_grad=True)

In [4]: y = x.to(torch.bfloat16)

In [5]: y
Out[5]: 
tensor([0., 0., 0., 0.], device='cuda:0', dtype=torch.bfloat16,
       grad_fn=<ToCopyBackward0>)

In [6]: y.grad_fn
Out[6]: <ToCopyBackward0 object at 0x7fb413b15120>

In [7]: x.requires_grad
Out[7]: True

In [8]: y.requires_grad
Out[8]: True
```

------

Before (`master` branch):

```python
In [1]: import os
   ...: os.environ.update(LOCAL_RANK='0', LOCAL_WORLD_SIZE='1', RANK='0', WORLD_SIZE='1', MASTER_ADDR='127.0.0.1', MASTER_PORT='29700')

In [2]: import deepspeed
   ...: import torch
   ...: import torch.distributed as dist
[2023-08-11 23:07:20,618] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)

In [3]: deepspeed.logger.setLevel('WARNING')

In [4]: with deepspeed.zero.Init(dtype=torch.bfloat16):
   ...:     print(torch.zeros(3))
   ...:     
tensor([0., 0., 0.], device='cuda:0', dtype=torch.bfloat16)

In [5]: with deepspeed.zero.Init(dtype=torch.bfloat16):
   ...:     print(torch.zeros(3, requires_grad=True))
   ...:     
tensor([0., 0., 0.], device='cuda:0', dtype=torch.bfloat16,
       grad_fn=<ToCopyBackward0>)
```

After (this PR):

```python
In [1]: import os
   ...: os.environ.update(LOCAL_RANK='0', LOCAL_WORLD_SIZE='1', RANK='0', WORLD_SIZE='1', MASTER_ADDR='127.0.0.1', MASTER_PORT='29700')

In [2]: import deepspeed
   ...: import torch
   ...: import torch.distributed as dist
[2023-08-11 23:08:06,064] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)

In [3]: deepspeed.logger.setLevel('WARNING')

In [4]: with deepspeed.zero.Init(dtype=torch.bfloat16):
   ...:     print(torch.zeros(3))
   ...:     
tensor([0., 0., 0.], device='cuda:0', dtype=torch.bfloat16)

In [5]: with deepspeed.zero.Init(dtype=torch.bfloat16):
   ...:     print(torch.zeros(3, requires_grad=True))
   ...:     
tensor([0., 0., 0.], device='cuda:0', dtype=torch.bfloat16, requires_grad=True)
```

### CODE REVIEW

**Comment by XuehaiPan** [comment]
@tohtana Thanks for the suggestion. I have updated the patch accordingly.

**Review by tohtana — APPROVED** [review]
